src/instructlab/cli/model/upload.py

```diff
@@ -0,0 +1,245 @@
+        if os.path.exists(self.model):
+            self.local_model_path = Path(self.model)
+            self.model = self.model.split("/")[-1]
+        else:
+            self.local_model_path = os.path.join(self.model_dir, self.model)
+            # throw an error if the built path is invalid
+            if not os.path.exists(self.local_model_path):
+                click.secho(
+                    f"Couldn't find model at {self.local_model_path} - are you sure it exists?",
+                    fg="red",
+                )
+                raise click.exceptions.Exit(1)
+
+        # upload either safetensors or gguf, if valid
+        if is_model_safetensors(self.local_model_path):
+            self.upload_safetensors()
+        elif is_model_gguf(self.local_model_path):
+            self.upload_gguf()
+        else:
+            click.secho(
+                f"Local model path {self.local_model_path} is a valid path, but is neither safetensors nor a GGUF - cannot upload to Hugging Face",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+    def upload_gguf(self) -> None:
+        try:
+            hf_api = HfApi()
+            hf_api.upload_file(
+                path_or_fileobj=self.local_model_path,
+                path_in_repo=self.model,
+                repo_id=self.upload_dest,
+                revision=self.release,
+                token=self.hf_token,
+            )
+        except Exception as exc:
+            click.secho(
+                f"\nUploading GGUF model failed with the following Hugging Face Hub error:\n{DEFAULT_INDENT}{exc}",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+        click.secho(
+            "\nUploading GGUF model succeeded!",
+            fg="green",
+        )
+
+    def upload_safetensors(self) -> None:
+        try:
+            hf_api = HfApi()
+            hf_api.upload_folder(
+                folder_path=self.local_model_path,
+                repo_id=self.upload_dest,
+                revision=self.release,
+                token=self.hf_token,
+            )
+        except Exception as exc:
+            click.secho(
+                f"\nUploading safetensors model failed with the following Hugging Face Hub error:\n{DEFAULT_INDENT}{exc}",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+        click.secho(
+            "\nUploading safetensors model succeeded!",
+            fg="green",
+        )
+
+
+@click.command()
+@click.option(
+    "--model",
+    "-m",
+    required=True,
+    help="Local model to upload.",
+)
+@click.option(
+    "--model-dir",
+    default=lambda: DEFAULTS.CHECKPOINTS_DIR,
+    show_default="The default system model location store, located in the data directory.",
+    help="The local directory to upload the model files from.",
+)
+@click.option(
+    "--dest-type",
+    type=click.Choice(["hf", "oci", "s3"]),
```

Comment:
s3 and oci are going to require additional info of where to publish.  I am guessing this param will still make sense.  But why bake it into the API before the rest of the implementation is designed?

Reply:
I'm assigned to do OCI and S3 in the short-term following HF - I preferred skeletoning out some of that work for myself now

Reply:
It would be good to be certain -dest-type is still useful with oci and s3.  I assume we will also need a param for s3 url and some way to auth.  Similar for oci.  At that point dest-type would be repeat information.  Which may be fine, but worth thinking through before codifying.

Follow-on question of whether s3 implies AWS or any s3 compliant api.  And if it is any s3 compliant api, would the auth setup vary?

Reply:
Right now I'm planning on using `--dest-type` for all three - that may change in the future, but would prefer to keep it in for now.

Let's keep the convo around S3 here: https://github.com/instructlab/instructlab/issues/2671